### PR TITLE
Support disposing of EventEmitters

### LINKED ISSUE #2509: Support disposing of EventEmitters
VS Code disposes Events which frees references to all its listeners, we should do the same:

https://github.com/microsoft/vscode/blob/4d1e08fec954853878c692a4c6629d7f0f884fd0/src/vs/base/common/event.ts#L591-L602

### PR DESCRIPTION
**Pull request by revanthmahesh**
dispose function to clear all the listener references at once.

Fixes #2509

### CODE REVIEW

File: src/common/EventEmitter.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { IDisposable } from 'common/Types';
 
 interface IListener<T> {
   (e: T): void;
+  (): void;
```
Comment (Tyriar) [inline src/common/EventEmitter.ts]: Why was this added?

Comment (Tyriar) [reply src/common/EventEmitter.ts]: This is probably the cause of the build failure

Comment (revanthmahesh) [reply src/common/EventEmitter.ts]: IListener is allowing only a function with at least one param. It is not allowing Array.prototype.clear. In vs code it is  handled by type LinkedList.
